Paths: record-editor.py

swagneto opened the pull request:
Topic: Record Editor Qt Best Practices Refactor
Description:
- Made several refactors to the classes in record-editor.py to better
  comply with Qt best practices. Changes include:
  - All Qt5 objects now imported via wildcard for convenience.
  - RecordWindow now utilizes sizeHint to size itself implicitly.
  - Removed all function pointer callbacks and replaced them with
    Qt signals which accomplish the same.
  - Refactored QStandardItemModel population so new rows are appended
    rather than inserted.
  - Delegated setting of QMainWindow central widget to the inherited
    create_layout methods.
  - Moved the selection layout into a QDockWidget for easy repositioning
    in PhotoRecordViewer.
  - Removed all superfluous QGroupBox widgets from the various
    create_layout methods, and replaced setting of fixed size policies
    with the appropriate usage of stretch factors.
  - Discovered a bug that causes a hard crash when triggering the close
    menu action on a ArtRecordEditor. Commented out usage of the action
    for the time being (closing via X button does not cause crash).
  - Added pySlot decorators where necessary.
  - Added constants for model roles and column names.
  - Misc whitespace/typo/comment fixes.

sdcolli1 committed:
Paths: record-editor.py
Topic: Record Editor Qt Best Practices Refactor
Description:
  * Made several refactors to the classes in record-editor.py to better
    comply with Qt best practices. Changes include:
      - All Qt5 objects now imported via wildcard for convenience.
      - RecordWindow now utilizes sizeHint to size itself implicitly.
      - Removed all function pointer callbacks and replaced them with
        Qt signals which accomplish the same.
      - Delegated setting of QMainWindow central widget to the inherited
        create_layout methods.
      - Moved the selection layout into a QDockWidget for easy repositioning
        in PhotoRecordViewer.
      - Removed all superfluous QGroupBox widgets from the various
        create_layout methods, and replaced setting of fixed size policies
        with the appropriate usage of stretch factors.
      - Discovered a bug that causes a hard crash when triggering the close
        menu action on a ArtRecordEditor. Added a comment about its existence.
      - Added pySlot decorators where necessary.
      - Added constants for model roles and column names.
      - Misc whitespace/typo/comment fixes.

  * Refactored PhotoRecordEditor and ArtRectorEditor to utilize
    a pre-allocated QStandardItemModel, and a SelectionView class
    based off of QTableView. These changes ensure model initialization
    happens efficiently, and the selection views are properly sized
    to display the contents of all of their columns.

sdcolli1 committed:
Paths: record-editor.py
Topic: Record Editor Qt Best Practices Refactor
Description:
  * Made several refactors to the classes in record-editor.py to better
    comply with Qt best practices. Changes include:
      - All Qt5 objects now imported via wildcard for convenience.
      - RecordWindow now utilizes sizeHint to size itself implicitly.
      - Removed all function pointer callbacks and replaced them with
        Qt signals which accomplish the same.
      - Delegated setting of QMainWindow central widget to the inherited
        create_layout methods.
      - Moved the selection layout into a QDockWidget for easy repositioning
        in PhotoRecordViewer.
      - Removed all superfluous QGroupBox widgets from the various
        create_layout methods, and replaced setting of fixed size policies
        with the appropriate usage of stretch factors.
      - Discovered a bug that causes a hard crash when triggering the close
        menu action on a ArtRecordEditor. Added a comment about its existence.
      - Added pySlot decorators where necessary.
      - Added constants for model roles and column names.
      - Misc whitespace/typo/comment fixes.

  * Refactored PhotoRecordEditor and ArtRectorEditor to utilize
    a pre-allocated QStandardItemModel, and a SelectionView class
    based off of QTableView. These changes ensure model initialization
    happens efficiently, and the selection views are properly sized
    to display the contents of all of their columns.

ghost commented:
Hey man! Hope I did this right, figured this would be a commit to get the ball rolling on github for code-related discussions. Have at it.

gthomsen commented on record-editor.py:
```diff
@@ -265,30 +264,29 @@ def create_models( self ):
+            filename_item.setData( photo["filename"] )
+            state_item.setData( photo["state"] )
+
+            self.photosModel.appendRow( [id_item, filename_item, state_item] )
```
It looks like this line is the culprit for a huge increase in startup time.  With a database containing ~4100 records startup time went from ~2 seconds (mostly from image loading) to ~11 seconds.  I vaguely remember reading something about performance of append vs insert and having to do something (freezing the model?) to eliminate difference between the two.

swagneto replied:
I have an idea that may help ameliorate this. Can you apply the patch I've attached and let me know how it performs against your database instance? I was able to see a modest improvement on my build (init time went from ~0.45 secs to ~0.3), but only against a database.xml with about 1000 photos, and most of which I don't have the actual photos downloaded. Also, if this does help performance (and you don't have an issue using a list view instead of a tree view), whats the best way to roll it in? Can this pull request be amended, or should the refactor happen in a separate request?

[model_patch.txt](https://github.com/gthomsen/streetart-analysis/files/429164/model_patch.txt)

gthomsen replied:
That definitely brings it back to a sane startup time.  It is still slower
on my system but only by 5-10% (1.12s v 1.22s).  One noticeable thing is
that it's consistent in startup time while my branch has a 10% swing over
repeated runs.  Additionally, this also addresses another issue that I
didn't initially report in that clicking on the selection view's header to
sort records would cause a huge slow down as well.  Given that went away,
the startup isn't bad, and the initial sort matches the database's record
order I'll call this a win.

Definitely update this pull request with any changes made to address our
discussion.  Once everything is square I'll merge it.  Don't forget to
match the (attempted) coding style por favor.

On Sun, Aug 21, 2016 at 6:57 PM, swagneto notifications@github.com wrote:

> In record-editor.py
> https://github.com/gthomsen/streetart-analysis/pull/1#discussion_r75606349
> :
> 
> > ```
> >      for photo in self.photos:
> > ```
> > -            self.photosModel.insertRow( 0 )
> > -            self.photosModel.setData( self.photosModel.index( 0, 0 ), photo["id"] )
> > -            self.photosModel.setData( self.photosModel.index( 0, 1 ), photo["filename"] )
> > -            self.photosModel.setData( self.photosModel.index( 0, 2 ), photo["state"] )
> > -            id_item       = QStandardItem( photo["id"] )
> > -            filename_item = QStandardItem( photo["filename"] )
> > -            state_item    = QStandardItem( photo["state"] )
> >   +
> > -            id_item.setData( int( photo["id"] ) )
> > -            filename_item.setData( photo["filename"] )
> > -            state_item.setData( photo["state"] )
> >   +
> > -            self.photosModel.appendRow( [id_item, filename_item, state_item] )
> 
> I have an idea that may help ameliorate this. Can you apply the patch I've
> attached and let me know how it performs against your database instance? I
> was able to see a modest improvement on my build (init time went from ~0.45
> secs to ~0.3), but only against a database.xml with about 1000 photos, and
> most of which I don't have the actual photos downloaded. Also, if this does
> help performance (and you don't have an issue using a list view instead of
> a tree view), whats the best way to roll it in? Can this pull request be
> amended, or should the refactor happen in a separate request?
> 
> model_patch.txt
> https://github.com/gthomsen/streetart-analysis/files/429164/model_patch.txt
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> https://github.com/gthomsen/streetart-analysis/pull/1/files/41ae18f71c9958a8c58e39e1eab98c472e99a783#r75606349,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AJjQY1kE3XpJz8i4-RPmb6-ms4dfB14Kks5qiOX-gaJpZM4JmEfF
> .

gthomsen commented on record-editor.py:
```diff
@@ -533,30 +526,55 @@ def get_photo_id_from_selection( self ):
+                                                photo_id,
                                                 1,
-                                                Qt.MatchFixedString )[0]
+                                                Qt.MatchExactly )[0]
```
What's the motivation for doing QVariant-based matching vs string-based if the ID column is a string?

swagneto replied:
Note that DATA_ROLE is now being provided, which means the QVariant storing the ID integer is examined (which got set via QStandardItem.setData() in setup_model). I assumed this would be faster and more "correct" than string matching against the displayed (despite the ID column being hidden) text of the view.

gthomsen replied:
I had assumed they were linked though forgot to ask about that.  Good call
on the string vs integer matching - I had forgot about the data type
difference.

On Sun, Aug 21, 2016 at 7:04 PM, swagneto notifications@github.com wrote:

> In record-editor.py
> https://github.com/gthomsen/streetart-analysis/pull/1#discussion_r75606461
> :
> 
> > ```
> >                                              1,
> > ```
> > -                                                Qt.MatchFixedString )[0]
> > -                                                Qt.MatchExactly )[0]
> 
> Note that DATA_ROLE is now being provided, which means the QVariant
> storing the ID integer is examined (which got set via
> QStandardItem.setData() in setup_model). I assumed this would be faster and
> more "correct" than string matching against the displayed (despite the ID
> column being hidden) text of the view.
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> https://github.com/gthomsen/streetart-analysis/pull/1/files/41ae18f71c9958a8c58e39e1eab98c472e99a783#r75606461,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AJjQY98B8g2Q29_9CWHJUurp5QE43fkSks5qiOdygaJpZM4JmEfF
> .

gthomsen commented on record-editor.py:
```diff
@@ -813,22 +834,26 @@ def create_models( self ):
+            id_item.setData( art["id"] )
+            state_item.setData( art["state"] )
+
+            self.artModel.appendRow( [id_item, state_item] )
```
Whatever happens with the PhotoRecordViewer's photoModel above should be done here as well.  There won't ever be enough records to worry about performance though consistency is key.

swagneto replied:
Agree, once we have consensus on best approach (roll back to inserts, preallocate rows and set, etc...), I'll propagate to other models.

gthomsen commented on record-editor.py:
```diff
@@ -1191,9 +1191,13 @@ def create_new_record( self ):
+        id_item.setData( new_art_record["id"] )
+        state_item.setData( new_art_record["state"] )
+
+        self.artModel.appendRow( [id_item, state_item] )
```
Similar comment about consistency on insert vs append.

gthomsen commented on record-editor.py:
```diff
@@ -1691,40 +1721,42 @@ def create_layout( self ):
+        # a hard crash. For now the user will have to resort to using
+        # the X button, which does not cause the crash.
+        # self.closeAct  = QAction( "C&lose", self, shortcut="Ctrl+W",
+        #                           triggered=self.close )
```
Not sure where this crash comes from.  On an Ubuntu 14.04.4 LTS system, a Fedora 20, and a Fedora 24 system the keyboard shortcut works without issue and has for some 4500 record entries (~4000 on the Ubuntu system, ~500 on the Fedora systems).  Is it present in the original code before the other changes in this commit?

swagneto replied:
I'm able to reproduce on your unmodified master branch, but the key thing is you have to **click** the Close Window action, not use the shortcut (which seems to work just fine). I get this back from pycham after the hard crash:

Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)

I don't want to upset your workflow so I'm fine rolling back this change (can it be done from within this pull request?). Better fix is to probably avoid having Photo/ArtRecordEditor inherit QMainWindow and make them modal QWidgets instead, but that would be pretty low priority fix I would imagine.

gthomsen replied:
Okay, now I see what you're talking about - activating the menu nukes it.
Apparently I can't read.

I'm good with leaving this in and then indirectly addressing it by
migrating to modal widgets.  That would fall inline with code I'd like to
commit that pulls in the *RecordEditor classes for a supplemental editor
embedded within another application (in this case a viewer).

Go ahead and rollback this change within this PR.  Do you want to tackle
migrating the inheritance issue in this PR or set it aside for a future
PR?  Either way it shouldn't get in the way of what I'm doing.

On Sun, Aug 21, 2016 at 7:31 PM, swagneto notifications@github.com wrote:

> In record-editor.py
> https://github.com/gthomsen/streetart-analysis/pull/1#discussion_r75607067
> :
> 
> > ```
> >      Takes no arguments.
> > 
> >      Returns nothing.
> >      """
> > ```
> > 
> > ## 
> > -        self.closeAct  = QAction( "C&lose", self, shortcut="Ctrl+W",
> > -                                  triggered=self.close )
> > -        # XXX: for whatever reason, closing the window via menu action
> > -        # when there are more than two QMainWindow instances causes
> > -        # a hard crash. For now the user will have to resort to using
> > -        # the X button, which does not cause the crash.
> > -        # self.closeAct  = QAction( "C&lose", self, shortcut="Ctrl+W",
> > -        #                           triggered=self.close )
> 
> I'm able to reproduce on your unmodified master branch, but the key thing
> is you have to _click_ the Close Window action, not use the shortcut
> (which seems to work just fine). I get this back from pycham after the hard
> crash:
> 
> Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)
> 
> I don't want to upset your workflow so I'm fine rolling back this change
> (can it be done from within this pull request?). Better fix is to probably
> avoid having Photo/ArtRecordEditor inherit QMainWindow and make them modal
> QWidgets instead, but that would be pretty low priority fix I would imagine.
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> https://github.com/gthomsen/streetart-analysis/pull/1/files/41ae18f71c9958a8c58e39e1eab98c472e99a783#r75607067,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AJjQY3OBAHDBOnbJxcRkbivwYQbFTWZUks5qiO32gaJpZM4JmEfF
> .

gthomsen commented on record-editor.py:
```diff
@@ -361,6 +358,12 @@ def create_widgets( self ):
 
+        # dock widget which will hold the selection layout once created
+        # in create_layout, for now it gets an empty widget.
+        self.selection_dock = QDockWidget()
```
Can you educate me on the sizing strategy behind adding a dock?  From what I can tell it makes it easier to stretch the selection layout's size to be fully visible (and beyond, strangely) though don't understand if there are other benefits for resizing the PhotoRecordViewer window while maintaining preview aspect ratio, sensibly packed labels, etc.

swagneto replied:
The idea was to set the layout stretch factor(s) such that the scaling image portion of the window gets all extra space to expand (or contract) from resizing of the window, yet still have a way to easily resize the selection view if necessary. You also just get some niceties for free with dock widgets, like being able to reposition dock widgets to any of the available dock areas (top, bottom, left, right), or have them float outside the main window (though I disabled this since I didn't think it necessary). Note that this was not intended to solve the problem of the selection view not being fully visible with regards to is column sizes at initialization. I think that will take creating a subclass of whatever View class to be able to override sizeHint().

gthomsen replied:
Gotcha.  Can we roll the *View subclassing into this pull request to get
the sizing?

On Sun, Aug 21, 2016 at 7:12 PM, swagneto notifications@github.com wrote:

> In record-editor.py
> https://github.com/gthomsen/streetart-analysis/pull/1#discussion_r75606680
> :
> 
> > @@ -361,6 +358,12 @@ def create_widgets( self ):
> >          self.infoTakenLabel    = QLabel()
> >          self.infoTagsLabel     = QLabel()
> > -        # dock widget which will hold the selection layout once created
> > -        # in create_layout, for now it gets an empty widget.
> > -        self.selection_dock = QDockWidget()
> 
> The idea was to set the layout stretch factor(s) such that the scaling
> image portion of the window gets all extra space to expand (or contract)
> from resizing of the window, yet still have a way to easily resize the
> selection view if necessary. You also just get some niceties for free with
> dock widgets, like being able to reposition dock widgets to any of the
> available dock areas (top, bottom, left, right), or have them float outside
> the main window (though I disabled this since I didn't think it necessary).
> Note that this was not intended to solve the problem of the selection view
> not being fully visible with regards to is column sizes at initialization.
> I think that will take creating a subclass of whatever View class to be
> able to override sizeHint().
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> https://github.com/gthomsen/streetart-analysis/pull/1/files/41ae18f71c9958a8c58e39e1eab98c472e99a783#r75606680,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AJjQY8IGgnI_wjpsSJvSTSS_D3MVbR53ks5qiOmFgaJpZM4JmEfF
> .

gthomsen commented:
Aside from the performance regression (appending rows in QStandardItemModels vs inserting) and the disabling of ArtRecordViewer's close keybinding these look great.

ghost commented:
Hey I think I've made all the changes I'm going to for this pull request (which have been squashed into the initial commit), so unless you seen anything else feel free to merge.

sdcolli1 committed:
Merge branch 'master' of https://github.com/swagneto/streetart-analysis

gthomsen commented on record-editor.py:
```diff
@@ -265,30 +321,31 @@ def create_models( self ):
         # processing state.
         self.proxyPhotosModel = QSortFilterProxyModel()
-        self.proxyPhotosModel.setFilterKeyColumn( 2 )
+        self.proxyPhotosModel.setFilterKeyColumn( self.PATH_COLUMN )
```
This should be self.STATE_COLUMN, otherwise the combo box's activation will filter to an empty selection for all entries except "all".

gthomsen commented on record-editor.py:
```diff
@@ -212,15 +195,88 @@ def commit_record( self ):
+                          for i in range( self.model().columnCount() )] )
+
+            if self.verticalScrollBar().isVisible():
+                width += self.verticalScrollBar().width()
```
If I add 12 or more to this value then the horizontal scroll bar disappears due to sizeHint() matching the actual width.  A cursory search doesn't enlighten me as to why as what you have here looks correct.  Does the horizontal scroll bar appear by default for you?

gthomsen commented:
Can you confirm you rebased and pushed changes?  I ended up pulling with --rebase to get the update without an issue though I'm not sure if that was actually required to look at things locally.  Aside from the last two (hopefully) minor things I noticed, this looks great.

sdcolli1 committed:
Paths: record-editor.py
Topic: Record Editor Refactor PR Fixes
Description:
  * Fixed a bug where the wrong column was specified as the filter
    key column for PhotoRecordViewer's filter proxy model.
  * Adjusted the size hint for SelectionView so that the width
    returned is just large enough such that no horizontal scroll
    bar is necessary.
  * Adjusted the setting of the default row height to be based on
    the height of the display font.

ghost commented:
Alright, think I've addressed your last couple comments. In the end I couldn't quite figure out how to get the SelectionView.sizeHint() to return the appropriate width completely programatically, at least not when auto stretch is enabled for the last column. I can revisit this at some point, but I'm planning to move onto improving the photo preview area(s), then prototyping the widgets you requested.

gthomsen commented:
Merged.  Thanks for the revisions!

The sizing on my laptop is still off (3 pixels this time) though I think that will be a game of whack-a-mole where figuring out what is missing from the calculation is more important than finding a universal fudge factor.
